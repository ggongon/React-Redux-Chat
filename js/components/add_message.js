import React, {PropTypes} from 'react';

export default class AddMessage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      messageText: ""
    }
    this.onChange = this.onChange.bind(this);
  }

  onChange(e) {
    this.setState({messageText: e.target.value});
  }
  render() {
    return (
      <footer className="footer">
        <form className="flex row">
          <input className="stretchy" type="text" placeholder="Type a message and hit enter..." value={this.state.messageText} onChange={this.onChange} />
          <button type="submit">Add</button>
        </form>
      </footer>
    );
  }
}
