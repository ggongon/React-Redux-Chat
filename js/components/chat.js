import React, {PropTypes} from 'react';
import RoomList from './room_list'
import Room from './room';

export default (props) => {
    return (
      <div id="app" className="flex row">
        <div id="left">
          <RoomList />
        </div>
        <div id="right">
          <Room />
        </div>
      </div>
    );
}
