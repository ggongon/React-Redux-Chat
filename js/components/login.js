import React, {Component} from 'react';
import {connect} from 'react-redux';
import * as actions from '../actions';


export class Login extends Component {

  loginUser(event) {
    event.preventDefault();
    // update the state.username
    this.props.authenticate(this.refs.login.value);
  }

  test() {
    this.props.spy && this.props.spy();
  }

  componentWillUpdate(nextProps) {
    if(nextProps.username) {
      this.context.router.history.push("/chat");
    }
  }

  render() {
    return (
      <div id="loginContainer" className="flex absCenter">
        <form  className="loginContent flex absCenter col" onSubmit={this.loginUser.bind(this)}>
          <input type="text" ref="login" name="login" placeholder="Username"/>
          {/* <input type="password" ref="password" name="password" placeholder="Password"/> */}
          <button type="submit" name="submit" onClick={this.test.bind(this)}>Chat now!</button>
        </form>
      </div>
    );
  }
}

Login.contextTypes = {
  router: React.PropTypes.object
}

function mapStateToProps({username}) {
  return {username}
}

export default connect(mapStateToProps, actions)(Login);
