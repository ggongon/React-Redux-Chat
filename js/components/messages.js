import React, {Component} from 'react';
import {connect} from 'react-redux';
import * as actions from '../actions'

export class Messages extends Component {
  constructor(props) {
    super(props);
  }

  componentWillMount () {
    if (this.props.currentRoomId === null) {
      console.log("cwm:", this.props.currentRoomId)
      this.props.fetchMessages(this.props.currentRoomId);
    }
  }

  componentWillUpdate(nextProps) {
    console.log("cwu:", nextProps.currentRoomId)
    if (nextProps.currentRoomId !== this.props.currentRoomId) {
      this.props.fetchMessages(nextProps.currentRoomId);
    }
  }

  renderMessages() {
    if (!this.props.messages) return;

    return this.props.messages.map((message, index) => {
      return (
        <li key={index}>
          <div className="message">{message.message}</div>
          <div className="user">{message.name}</div>
        </li>
      )
    })
  }

  render() {
    console.log(this.props.currentRoomId, this.props.messages)
    if (!this.props.messages.length) return <div className="stretchy">No messages.</div>

    return (
      <div id="messages" className="stretchy">
        <ul>
          {this.renderMessages()}
        </ul>
      </div>
    );
  }
}

const mapStateToProps = ({currentRoomId, messages}) => {
  return {
    currentRoomId, messages
  }
}

export default connect(mapStateToProps, actions)(Messages);
