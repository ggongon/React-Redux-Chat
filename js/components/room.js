import React, {Component} from 'react';
import {connect} from 'react-redux';
import RoomInfo from './room_info';
import Messages from './messages';
import AddMessage from './add_message';

export class Room extends Component {
  render() {
    if (this.props.currentRoomId === null) {
      return <div className="flex stretchy absCenter">Choose a room.</div>
    }
    return (
      <div ref="room" className="flex col">
        <RoomInfo />
        <Messages />
        <AddMessage />
      </div>
    );
  }
}

const mapStateToProps = ({currentRoomId}) => {
  return {currentRoomId}
}

export default connect(mapStateToProps)(Room);
