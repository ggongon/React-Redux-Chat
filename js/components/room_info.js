import React, {Component} from 'react';
import {connect} from 'react-redux';
/**
 * sample response:
 * const currentRoom =  {
     name: 'Tea Chats',
     id: 0,
     users: [ 'Ryan', 'Nick' ]
   }
 */

export class RoomInfo extends Component {
  constructor(props) {
    super(props);
  }

  render() {
    let currentRoom = this.props.currentRoom;

    if (!Object.keys(currentRoom).length) return <div>get a room</div>
    let users = currentRoom.users.join(", ");
    return (
      <div className="roomInfo center">
        <h2>{currentRoom.name}</h2>
        <h5>{users}</h5>
      </div>
    );
  }
}

const mapStateToProps = ({currentRoom}) => {
  return {currentRoom}
};

export default connect(mapStateToProps)(RoomInfo);
