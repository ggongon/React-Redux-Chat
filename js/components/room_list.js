import React, {PropTypes} from 'react';
import {connect} from 'react-redux';
import * as actions from '../actions';

export class RoomList extends React.Component {

  constructor(props) {
    super(props);
    this.renderRooms = this.renderRooms.bind(this);
    this.onRoomClicked = this.onRoomClicked.bind(this);
  }

  componentWillMount() {
    this.props.fetchRooms();
  }

  onRoomClicked(roomId) {
    this.props.spy && this.props.spy();
    if (roomId !== this.props.currentRoomId) {
      this.props.setCurrentRoomId(roomId);
      this.props.fetchRoom(roomId);
    }
  }

  renderRooms() {
    if (this.props.rooms) {
      return this.props.rooms.map((room) => {
        return (
          <li key={room.id} onClick={() => this.onRoomClicked(room.id) }>
            {room.name}
          </li>
        )
      })
    }
  }

  render() {
    return (
      <div>
        <ul>
          {this.renderRooms(this.props.rooms)}
        </ul>
      </div>
    )
  }
}

// RoomList.PropTypes = {
//   rooms: PropTypes.array,
// }

function mapStateToProps({rooms, currentRoomId}) {
  return { rooms, currentRoomId }
}

export default connect(mapStateToProps, actions)(RoomList);
