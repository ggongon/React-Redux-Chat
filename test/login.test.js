import React from 'react';
import renderer from 'react-test-renderer';
import {mount, shallow} from 'enzyme'; // for testing actual component content
import sinon from 'sinon';
import {Login} from '../js/components/login';

test("should render a component", () => {
  const login = renderer.create(<Login />)
  expect(login).toMatchSnapshot();
})

test("should have all the expected login form elements", () => {
  let login = mount(<Login />);

  // username
  expect(login.find("[name='login']").length).toBe(1);
  // password
  expect(login.find("[name='password']").length).toBe(1);
  // login button
  expect(login.find("[name='submit']").length).toBe(1);
})

test("clicking submit should authenticate the user", () => {
  // passes login and password to action
  const spy = jest.fn();
  const props = {spy};
  const login = mount(<Login {...props} />);

  login.find('button').simulate('click');
  // action should update the state.authenticated to true
  expect(spy).toHaveBeenCalled();
})
