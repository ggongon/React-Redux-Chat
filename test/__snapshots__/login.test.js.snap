// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a component 1`] = `
<div
  className="flex absCenter"
  id="loginContainer"
>
  <form
    className="loginContent flex absCenter col"
    onSubmit={[Function]}
  >
    <input
      name="login"
      placeholder="Username"
      type="text"
    />
    <button
      name="submit"
      onClick={[Function]}
      type="submit"
    >
      Chat now!
    </button>
  </form>
</div>
`;
